apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClientImpl.SingleChannel.cs

```diff
@@ -392,13 +401,34 @@ private void HandlePullMessageData(Task<bool> moveNextTask)
                 receivedMessages.Clear();
                 // Get all ack-ids, used to extend leases as required.
                 var msgIds = new HashSet<string>(msgs.Select(x => x.AckId));
+                _receiptModAckForExactlyOnceDelivery = _exactlyOnceDeliveryEnabled;
+                if (_exactlyOnceDeliveryEnabled)
+                {
+                    // Populate receipt ModAck dictionary before the lease extends are send.
+                    PopulateAckIdsForReceiptModAck(msgIds);
+                }
+
                 // Send an initial "lease-extension"; which starts the server timer.
                 HandleExtendLease(msgIds, null);
-                // Asynchonously start message processing. Handles flow, and calls the user-supplied message handler.
-                // Uses Task.Run(), so not to clog up this "master" thread with per-message processing.
-                Task messagesTask = _taskHelper.Run(() => ProcessPullMessagesAsync(msgs, msgIds));
-                // Once all received messages have been queued for processing, read the stream for more messages.
-                Add(messagesTask, Next(true, () => HandlePullMoveNext(null)));
+                if (_exactlyOnceDeliveryEnabled)
+                {
+                    // In the exactly-once delivery flow, we must send only those messages to the user handler which had successful receipt ModAck responses.
+                    // This is necessary to prevent message redelivery. We get the status of success or permanent failure instantly.
+                    // However, temporary failures are retried for up to three times and may eventaully succeed, result in permanent failure, or remain as temporary failure.
+                    // Therefore, we must wait for all receipt ModAck responses to complete to obtain the final status.
+                    // Then, the messages with successful receipt ModAcks are sent to the user, while those with failed ModAcks are removed from further processing.
+                    Add(_eventReceiptModAckForExactlyOnceDelivery.WaitAsync(_softStopCts.Token)
+                        .ContinueWith(task => ProcessSuccessfulMessages(msgs, msgIds), CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, _taskHelper.TaskScheduler)
+                        , Next(true, () => HandlePullMoveNext(null)));
```

Comment:
What is HandlePullMoveNext here? I suspect we may want to rename it for clarity.